mindmaps-graql/src/main/java/io/mindmaps/graql/Reasoner.java

```diff
@@ -440,56 +442,86 @@ private Query applyRuleToAtom(Atomic parentAtom, Rule child, Map<String, Type> v
         return expansions;
     }
 
-    private Set<Set<Substitution>> DBlookup(Query query)
+    private Set<Map<String, Concept>> DBlookup(Query query)
     {
-        Set<Set<Substitution>> answers = new HashSet<>();
+        Set<Map<String, Concept>> answers = new HashSet<>();
 
-        List<Map<String, Concept>> results = Lists.newArrayList(query.getMatchQuery());
+        List<Map<String, Concept>> results = Lists.newArrayList(query.getMatchQuery().distinct());
```

Comment:
Iterate directly over the stream into the set rather than putting it in a list and then into the set.